### ISSUE TITLE
Issues with standard ranged tags specification

### ISSUE THREAD

**Issue by gzagatti** [nvim-neorg/neorg#577]
The [current specification 0.1](https://github.com/nvim-neorg/norg-specs/blob/main/0.1-specification.norg) defines the `Standard Ranged Tag` as isolated `norg` blocks within the document.

My main concern is that it is not clear how the struture of the text changes when these ranged tags are introduced. In fact the current implementation of `norg` treats the text within ranged tags as part of the main text. 

For instance, in the code below:

```norg
* Heading 1

Some text that leads to an example.

|example
* Example heading

The example goes like this.
  - item 1
  - item 2
|end
```

The current `norg` implementation will treat `Example heading` as `Heading 1` in the main text. Currently, the text will fold at `* Heading 1` and `* Example heading` which makes it difficult to read since semantically the standard ranged tag should really belong to `* Heading 1` and be folded under it, not besides it.

The specification mentions that the example tag "is a simple way to show an example of some Norg markup without it being rendered and treated as physicial markup". In this case, I feel that a verbatim tag like `@example` would be more appropriate as it does not change the structure of the main text and places the code block inside of the element to which it belongs. In this case, I would also argue that you want to demonstrate the use of the markup and not really render the markup.

I have similar concerns with `|comment`, `|detail` and `|group`. But, in these cases it could indeed be quite convenient to express ideas in parallel to the main text. It is however important that we determine whether these items should be treated as nested inside of the parent element or as a sibling to the nested element. I could imagine both situations. For instance, a textbook might contain textboxes to expand on an idea. The textbox might refer to an idea presented in the chapter or in the current heading. Perhaps a markup to distinguish these cases might be warranted.

I also feel that there might be some overlap between `Range-able Detached Modifiers` and `Standard Ranged Tags`. They both modify `norg` text contained within it and they both delimit a piece of `norg` text. We could even imagine something like:

```
|^ Ranged Footnote
Content of the footnote

Which scans up to the closing modifier.
|^
```

**Comment by vhyrro** [comment]
Also, to clarify a little bit. In the response, I said:
> Of course, not every parser or semantic analyzer is required to do this sort of crazy analysis, they can just hardcode certain behaviours for inbuilt and documented tags

I don't want to accidentally put you under the impression that a parser would have to do any level of semantic analysis, so i edited the comment to remove that bit. Norg's designed to be parseable left to right in a single pass with as little ambiguities as possible. Nobody said that the semantics will be simple, however ;), it still wants to be a powerful format!

**Comment by gzagatti** [comment]
Thanks for the thorough answer. It now makes much more sense what you trying to do with Standard Ranged Tags. 

Macros are great for a markup language and if you are looking for more inspiration check [Pollen](https://docs.racket-lang.org/pollen/)

I have a few more comments on your response.

If Standard Ranged Tags enter macro evaluation territory, why don't we consider unifying Macro definitions with Standard Ranged Tags such that they are in essence the same type of objects. We could then evaluate a Standard Ranged Tag with the Macro tag. Something like this:

```norg
|example Foo
* My example

This is an example with `norg` syntax
|end

=example Foo
```

Of course, you would not like to repeat an example twice. But who knows? Examples could even have parameters since they are macros after all. Thus, if I am writing multiple examples where only a few things change, I could use this feature.

The syntax could easily incorporate macro definitions.

```norg
|macro important

* Heading in macro

#color red
|group
  &...&
|end
|end

== important
This is some highly important markup

* And this is a very important heading
==
```

However, in this case I would invert the definition of the specification. Such that the syntax `|macro <name>` defines the macro and the syntax `== <name>` evaluates it. This change would actually agree with the the macro tag `=<name>`. In that the `=` signs always denotes evaluation. You could then even allow the following:

```norg
= important 
  This is some highly important markup
```

To evaluate the following line as important.

Also, since variables are macro in disguise with some additional constraints with what is allowed, you could just have then as macros and have inline evaluations using `=`:

```norg
|var foo
*this* is some /content/
|end

The content of the above variables is: =foo=
```

I believe that these modifications would bring substantial simplification to the language specification.

One of the issues with the above is that you might want to keep the one-liners for macro and variable definition. One option would be the following:

```norg
||example
This is an example that spans multiple things:
- one thing
- another
||

|example
This is a one line example; don't be overly ambitious as it only applies to this paragraph.
```

Whether the specifications changes or not, it should be clarified in the specification whether all syntax structures are allowed in a macro definition. Ideally, this would be the case, because you do not want to be limited by what a macro can define.

Next, it is not clear whether the Standard Ranged Tag should have valid norg or if it varies on a case-by-case. For instance, the specification uses a lot of `|example` to denote invalid markup.

Let's say I have this invalid example:

```norg
|example

$$ Open definition
I am an open definition but I won't be closed
|end

This continues the main text but should not be part of the definition.
```

Whatever comes after the end could be evaluated as part of the definition. Unless it is clear that the markup inside of the Standard Ranged Tag does not propagate or only propagates in a few cases, it is difficult to parse this text. What happens when we use `|group` in this case? Does it propagate? And if so, is only the beginning of the definition part of the group?

**Comment by vhyrro** [comment]
Some interesting ideas and concerns here! When mrossinek gets back I'll have a chat with him about this so we can establish exactly what behaviour we want. I'll let you know when him and I chat it out so we can discuss further :)

**Comment by mrossinek** [comment]
@vhyrro and I had a chat about your reply the other day and I finally had the time to proof-read the spec as well as go over this thread here in some more detail. Vhyrro already replied to your initial comment and I think that has answered all your questions. As per the follow-up, let me try to explain why we have decided to keep things as they are now.

<hr/>

I see your point about trying to simplify the language and while the idea of combining macros and standard ranged tags is intriguing, I believe that this will not be possible in general. Let me try to break this down a bit:

Macros are a very high-level aspect of Norg and they basically require a deep-dive into the metaprogramming capabilities of Norg which have not been fleshed out 100% (yet). I consider the use of macros a super-Norg-user tool and would like to avoid having to "scare" users with the explanations of these when they are just getting started with Norg.
I actually moved their explanation _after_ the explanation of the other tag types in the spec earlier today, exactly for this reason.

Having said that, while the internal implementations which interpret ranged tags _can_ implement them as macros internally, that is by no means required. It is even unlikely, if you gradually build up a parser, since standard ranged tags are part of layer 4, whereas generalized macros are part of the final layer (5).

<hr/>
Let me now talk about a few specific tags.

I think, verbatim ranged tags are clear and there is not too much to discuss here. The `@`-tags effectively swallow everything until an `@end` is reached. This is important to avoid any potential conflicts in any potential language for code-block for example.

We quickly found, that these are however not sufficient for all potential tag-applications, because we may want to have tags with nested norg-syntax, and, more importantly, which get incorporated into the document. That's where the `|`-tags come into play.

The `|details` and `|group` tag are utilities to help you typeset very specific scenarios. Personally I think, that their contents should be considered part of the same document. That's why Vhyrro meant earlier, with you being able to link to things that you have inside of these tags.

Now, the `|example` tag (as you correctly pointed out) currently has some issues with the `main` branches of Neorg and the treesitter parser. It is not intentional, that headings contained within those, are part of the overall structure. That said, this `|example` tag, is likely only ever going to be used to self-document norg. That is exactly why we use it, instead of `@code norg`, since we want it to behave the same way that its contents would as a standalone file.
I hope this makes sense...

<hr/>

Now a few more concrete points with respect to your proposed macro changes:
- we cannot have `|macro` be macro definitions, since that would require the `|macro`-tag to define itself, which does not really work the way it is intended
- we do not want to introduce "double-char"-tags (like in your `||example` suggestion)

And maybe a final more general comment:
One could ask the question, what objects are tags and what objects are (range-able) detached modifiers. The answer to that would go somewhat like this:
- we have detached modifiers for those syntax elements, which we consider to occur most commonly
- we have tags to achieve certain "behavior" which those previous detached modifiers cannot fulfill
- since we want Norg to be largely context-free, we have very limited combination possibilities for character use between tags, detached and attached modifiers, before running into ambiguities
